### ISSUE TITLE
trino-main tests sometimes fail with OutOfMemoryError

### ISSUE THREAD

**Issue by ksobolew** [trinodb/trino#12627]
The tests seem to run fine, until this appears:

```
Terminating due to java.lang.OutOfMemoryError: Java heap space
Error: Process completed with exit code 3.
```

And this fails the build. No further logs are present.

Example builds:

https://github.com/trinodb/trino/runs/6685877352?check_suite_focus=true
https://github.com/trinodb/trino/runs/6610883813?check_suite_focus=true

In each case the last test executed is different.

**Comment by ksobolew** [comment]
Possibly related? #9879 (thanks @hashhar)

**Comment by findepi** [comment]
https://github.com/trinodb/trino/runs/7749050866?check_suite_focus=true

**Comment by ebyhr** [comment]
https://github.com/trinodb/trino/actions/runs/3494298332/jobs/5849968695

**Comment by findepi** [comment]
From @ebyhr link above


```
Error:  java.lang.OutOfMemoryError: Java heap space
Error:  Terminating due to java.lang.OutOfMemoryError: Java heap space
....
Error:  Crashed tests:
Error:  io.trino.operator.join.BenchmarkHashBuildAndJoinOperators
Error:  io.trino.util.BenchmarkPagesSort
```

These two classes don't take more than 200m to run (didn't check lower than that).

The unclear part is whether these are actually all running classes

```
$ cat github-log-file.txt | grep '\[BEFORE CLASS]'  | awk '{print $NF}' | grep -vFf <( cat github-log-file.txt | grep '\[AFTER CLASS]'  | awk '{print $NF}' )
io.trino.operator.output.TestPositionsAppender
io.trino.operator.join.BenchmarkHashBuildAndJoinOperators
io.trino.util.BenchmarkPagesSort
```

`TestPositionsAppender` hasn't been mentioned by surefire as crashed.

```
$ cat github-log-file.txt | grep TestPositionsAppender
2022-11-17T23:11:14.564-0600 ProgressLoggingListener [BEFORE CLASS] io.trino.operator.output.TestPositionsAppender
2022-11-17T23:11:14.578-0600 ProgressLoggingListener [TEST START]   io.trino.operator.output.TestPositionsAppender.testSliceRle
2022-11-17T23:11:14.599-0600 ProgressLoggingListener [TEST SUCCESS] io.trino.operator.output.TestPositionsAppender.testSliceRle; (took: 0.0 seconds)
```

It's weird that it was started and hasn't completed, but this is probably due to the use of `@Test(priority)`  on it. (https://github.com/trinodb/trino/pull/12927/files#r1026126807)

TL;DR the crashed tests are NOT the cause of this OOM. Other options

- background threads leak (like left behind query runners, or threads started by them)
- static fields leak
- test execution framework leak

@nineinchnick can I ask your help getting a heap dump from the CI?

**Comment by nineinchnick** [comment]
I added uploading any `*.hprof` files as workflow run artifacts in #14865, PTAL

**Comment by findepi** [comment]
https://github.com/trinodb/trino/actions/runs/3515601406/jobs/5892101018

```
Error:  java.lang.OutOfMemoryError: Java heap space
Error:  Terminating due to java.lang.OutOfMemoryError: Java heap space
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
Warning:  Corrupted STDOUT by directly writing to native stream in forked JVM 1. See FAQ web page and the dump file /home/runner/work/trino/trino/core/trino-main/target/surefire-reports/2022-11-21T16-57-04_001-jvmRun1.dumpstream
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  17:44 min
[INFO] Finished at: 2022-11-21T17:14:43Z
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M5:test (default-test) on project trino-main: There are test failures.
Error:  
Error:  Please refer to /home/runner/work/trino/trino/core/trino-main/target/surefire-reports for the individual test results.
Error:  Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
Error:  The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
Error:  Command was /bin/sh -c cd /home/runner/work/trino/trino/core/trino-main && /opt/hostedtoolcache/Java_Zulu_jdk/17.0.5-8/x64/bin/java -Duser.timezone=America/Bahia_Banderas -Duser.language=en -Duser.region=US -Dfile.encoding=UTF-8 -Xmx3g -Xms3g -XX:+ExitOnOutOfMemoryError -XX:+HeapDumpOnOutOfMemoryError -XX:-OmitStackTraceInFastThrow -jar /home/runner/work/trino/trino/core/trino-main/target/surefire/surefirebooter12401623012388028830.jar /home/runner/work/trino/trino/core/trino-main/target/surefire 2022-11-21T16-57-04_001-jvmRun1 surefire10004513[9925](https://github.com/trinodb/trino/actions/runs/3515601406/jobs/5892101018#step:6:9926)04223094tmp surefire_113260158471551277845tmp
Error:  Error occurred in starting fork, check output in log
Error:  Process Exit Code: 3
Error:  Crashed tests:
Error:  io.trino.operator.BenchmarkGroupByHashOnSimulatedData
Error:  io.trino.util.BenchmarkPagesSort
Error:  org.apache.maven.surefire.booter.SurefireBooterForkException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
Error:  Command was /bin/sh -c cd /home/runner/work/trino/trino/core/trino-main && /opt/hostedtoolcache/Java_Zulu_jdk/17.0.5-8/x64/bin/java -Duser.timezone=America/Bahia_Banderas -Duser.language=en -Duser.region=US -Dfile.encoding=UTF-8 -Xmx3g -Xms3g -XX:+ExitOnOutOfMemoryError -XX:+HeapDumpOnOutOfMemoryError -XX:-OmitStackTraceInFastThrow -jar /home/runner/work/trino/trino/core/trino-main/target/surefire/surefirebooter12401623012388028830.jar /home/runner/work/trino/trino/core/trino-main/target/surefire 2022-11-21T16-57-04_001-jvmRun1 surefire10004513992504223094tmp surefire_113260158471551277845tmp
Error:  Error occurred in starting fork, check output in log
Error:  Process Exit Code: 3
Error:  Crashed tests:
Error:  io.trino.operator.BenchmarkGroupByHashOnSimulatedData
Error:  io.trino.util.BenchmarkPagesSort
```

**Comment by findepi** [comment]
Here's my attempt to try find tests that hold on to resources: https://github.com/trinodb/trino/pull/15151

**Comment by findepi** [comment]
https://github.com/trinodb/trino/pull/15165 is merged now. Let's see whether is helps anything.

https://github.com/trinodb/trino/pull/15181 is also merged, so hopefully next time tests OOM we will have some data to work with.

**Comment by findepi** [comment]
> #15165 is merged now. Let's see whether is helps anything.

No new reports since then. Let me optimistically close this. Let's reopen if problem reoccurs.
